gh-133390: Support SQL keyword completion for sqlite3 CLI

tanloong opened the pull request:
This adds tab-completion for the [147 SQLite Keywords](https://sqlite.org/lang_keywords.html). A whitespace is appended to completion candidates for users' convenience and to mimic the behavior of [sqlite3](https://sqlite.org/cli.html) tool.

<!-- gh-issue-number: gh-133390 -->
* Issue: gh-133390
<!-- /gh-issue-number -->

tanloong committed:
Support basic completion for sqlite3 command-line interface

StanFromIreland commented on Lib/sqlite3/__main__.py:
```diff
@@ -62,6 +63,59 @@ def runsource(self, source, filename="<input>", symbol="single"):
+                "TRANSACTION", "TRIGGER", "UNBOUNDED", "UNION", "UNIQUE",
+                "UPDATE", "USING", "VACUUM", "VALUES", "VIEW", "VIRTUAL",
+                "WHEN", "WHERE", "WINDOW", "WITH", "WITHOUT"]
+    options = [c + " " for c in keywords if c.startswith(text.upper())]
```
Why, currently they are all uppercase?

tanloong replied:
The `sqlite3` tool gives upper case completion for even lower case user input, I thought that is standard and took that as a reference. If you think it's not good I can modify the code to adapt the case of completion according to user input.

Just added news entry, I will add whatsnew entry and tests soon.

StanFromIreland reviewed (commented):
This needs a whatsnew entry, (maybe a note in the docs?), a NEWS entry and tests.

This is quite limited completion so I guess it could work.

tanloong committed:
Add news entry

picnixz commented on Lib/sqlite3/__main__.py:
```diff
@@ -62,6 +63,59 @@ def runsource(self, source, filename="<input>", symbol="single"):
+    except IndexError:
+        return None
+
+@contextmanager
```
```suggestion
        return None


@contextmanager
```

picnixz commented on Lib/sqlite3/__main__.py:
```diff
@@ -62,6 +63,59 @@ def runsource(self, source, filename="<input>", symbol="single"):
+    finally:
+        readline.set_completer(old_completer)
+
 def main(*args):
```
```suggestion
    finally:
        readline.set_completer(old_completer)


 def main(*args):
```

picnixz commented on Misc/NEWS.d/next/Library/2025-05-05-03-14-08.gh-issue-133390.AuTggn.rst:
```diff
@@ -0,0 +1 @@
+Support completion for :mod:`sqlite3` command-line interface.
```
```suggestion
Support keyword completion for :mod:`sqlite3` command-line interface.
```

tanloong committed:
Move completion code to separate module

StanFromIreland commented on Lib/sqlite3/_completer.py:
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
+
```
```suggestion
```
It is not necessary here.

StanFromIreland commented on Lib/sqlite3/_completer.py:
```diff
@@ -0,0 +1,58 @@
+    try:
+        readline.set_completer(_complete)
+        if readline.backend == "editline":
+            # libedit uses "^I" instead of "tab"
```
Is this a good enough reason for us to use it too?

tanloong replied:
Oh, I referred to pdb's completion implementation and pdb uses it. There are a few [same code blocks](https://github.com/search?q=repo%3Apython%2Fcpython+%2Fbind+%5C%5EI+rl_complete%2F&type=code) using it in `Lib/`.

I searched a bit and it looks like `libedit` does not recognize the literal string `tab` as a key name, instead it expects tab characters to be represented as `^I` (https://stackoverflow.com/a/7116997/30446581).

StanFromIreland commented on Doc/whatsnew/3.14.rst:
```diff
@@ -2176,6 +2176,9 @@ sqlite3
   it will now raise a :exc:`sqlite3.ProgrammingError`.
   (Contributed by Erlend E. Aasland in :gh:`118928` and :gh:`101693`.)
 
+* Support keyword completion for :mod:`sqlite3` command-line interface.
```
```suggestion
* Support keyword completion in the :mod:`sqlite3` command-line interface.
```

tanloong committed:
Update Lib/sqlite3/_completer.py

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

tanloong committed:
Update Doc/whatsnew/3.14.rst

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

tanloong committed:
Add test

tanloong committed:
Move keyword list to module level

StanFromIreland commented on Doc/whatsnew/3.14.rst:
```diff
@@ -2176,6 +2176,9 @@ sqlite3
 
+* Support keyword completion in the :mod:`sqlite3` command-line interface.
+  (Contributed by Long Tan in :gh:`133393`.)
+
```
This won't make it in to `3.14` I'm afraid.

StanFromIreland commented on Lib/sqlite3/_completer.py:
```diff
@@ -0,0 +1,58 @@
+
+
+def _complete(text, state):
+    options = [c + " " for c in _keywords if c.startswith(text.upper())]
```
```suggestion
    options = [c + " " for c in _keywords]
```

StanFromIreland replied:
This should be moved out of the functions, it will be unnecessarily regenerated.

tanloong replied:
The `if c.startswith()` filters out keywords that don't match prefix `text`. The filter is necessary otherwise user will get all keywords as suggestion to complete any input, and it can't be done outside of function where user's typed input `text` is not available.

tanloong replied:
OK I found how `_complete(text, state)` [should actually work](http://tiswww.case.edu/php/chet/readline/readline.html#How-Completing-Works):

> *state* is zero the first time the function is called, allowing the generator to perform any necessary initialization, and a positive non-zero integer for each subsequent call. Usually the generator function computes the list of possible completions when *state* is zero, and returns them one at a time on subsequent calls.

I modified the code to store candidates when state is 0. Regeneration is now avoided.

StanFromIreland reviewed (commented):
With https://github.com/python/cpython/issues/133447 in mind the keywords list would be quite handy (I think hardcoding is fine, it doesn't change often anyway). Maybe it could be moved elsewhere, what do you think Benedikt?

tanloong committed:
Remove whatsnew entry from 3.14

tanloong committed:
Avoid regeneration of candidates. Store them when state is 0 and returns
them one by one on subsequent calls

http://tiswww.case.edu/php/chet/readline/readline.html#How-Completing-Works

> *state* is zero the first time the function is called, allowing the
> generator to perform any necessary initialization, and a positive
> non-zero integer for each subsequent call. Usually the generator
> function computes the list of possible completions when *state* is zero,
> and returns them one at a time on subsequent calls.

tanloong committed:
Add whatsnew entry to 3.15

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +204,37 @@ def test_color(self):
 
+
+@requires_subprocess()
+class Completer(unittest.TestCase):
```
```suggestion
class CompletionTest(unittest.TestCase):
```

erlend-aasland replied:
`test_cli.py` does not use this class naming pattern. Let's follow the existing style.

erlend-aasland replied:
I reverted to a similar naming as the original change.

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -158,6 +163,45 @@ def test_interact_on_disk_file(self):
+    @classmethod
+    def setUpClass(cls):
+        # Ensure that the readline module is loaded
+        # If this fails, the test is skipped because SkipTest will be raised
```
```suggestion
```

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +204,36 @@ def test_color(self):
+        output = run_pty(script, input)
+        output = re.sub(rb"\x1b\[[0-9;]*[mK]", b"", output)
+        self.assertIn(b"SELECT", output)
+        self.assertIn(b"(1,)", output)
+
```
```suggestion
        self.assertIn(b"(1,)", output)


```

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +204,36 @@ def test_color(self):
+        # Then add 'EL' and complete 'SEL' to 'SELECT'.  Quit console in the end
+        # to let run_pty() return.
+        input = b"S\t\tEL\t 1;\n.quit\n"
+        output = run_pty(script, input)
```
Pass `env={'NO_COLOR': '1'}` to `run_pty` to disable colors.

StanFromIreland replied:
We have @force_not_colorized in support?

tanloong replied:
I tried `env={"NO_COLOR":"1"}` when I wrote this test, but still get colored output on my machine (Arch Linux). I tried again just now and no luck.

![图片](https://github.com/user-attachments/assets/0b78b32b-25e2-4af3-8734-eb47088768ed)

picnixz replied:
Mmh, something is wrong with the colorization? and I'm not sure `force_not_colorized` is sufficient as we're running in a subprocess.

tanloong replied:
I notice that [test_pdb.py](https://github.com/python/cpython/blob/main/Lib/test/test_pdb.py#L4770) passes CI even without either passing `env` or handling the control sequence in `output` of `run_pty()`.

I added `env={"NO_COLOR":"1"}` and removed `output = re.sub(rb"\x1b\[[0-9;]*[mK]", b"", output)` to see if CI passes, if it fails might need to add the `re.sub()` line back.

erlend-aasland replied:
This seems to have been fixed. Marking as resolved.

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +204,36 @@ def test_color(self):
+        script = "from sqlite3.__main__ import main; main()"
+        # List candidates starting with 'S', there should be multiple matches.
+        # Then add 'EL' and complete 'SEL' to 'SELECT'.  Quit console in the end
+        # to let run_pty() return.
```
The `input` is self-explanatory, no need to restate what the code does.

erlend-aasland replied:
Addressed; marking as resolved.

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +204,36 @@ def test_color(self):
+        if readline.backend == "editline":
+            raise unittest.SkipTest("libedit readline is not supported")
+
+    def test_keyword_completion(self):
```
Ideally, let's test when there is nothing to complete and in which order they are rendered. Make separate tests for that.

picnixz commented on Lib/sqlite3/_completer.py:
```diff
@@ -0,0 +1,61 @@
+from contextlib import contextmanager
+
+
+_keywords = ("ABORT", "ACTION", "ADD", "AFTER", "ALL", "ALTER", "ALWAYS",
```
We're anyway in a private module, so name this variable `KEYWORDS`. Ideally, I really want it coming from C but we can do it in a follow-up PR.

tanloong replied:
I moved `KEYWORDS` to C in the latest commit. I am not confident with my C code and look forward to your review. Thank you very much!

erlend-aasland replied:
Outdated; marking as resolved.

tanloong committed:
Address Bénédikt's review

tanloong committed:
Remove color handling of output; If CI fails might need to add back

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -235,5 +232,24 @@ def test_keyword_completion(self):
+        script = "from sqlite3.__main__ import main; main()"
+        input = b"zzzz\t;\n.quit\n"
+        output = run_pty(script, input, env={"NO_COLOR": "1"})
+        output = re.sub(rb"\x1b\[[0-9;]*[mK]", b"", output)
```
Let's find a way to avoid this. If this is a bug in `run_pty`, we should fix it (do we have some `FORCE_NO_COLOR` envvar perhaps?)

cc @hugovk as the colorization expert

StanFromIreland replied:
See: https://github.com/python/cpython/pull/133393#discussion_r2083226836

tanloong replied:
I figure it out. `run_pty()` has no bug. The problem is with GNU Readline, which [needs special settings to turn off color](https://www.gnu.org/software/bash/manual/html_node/Readline-Init-File-Syntax.html#index-colored_002dcompletion_002dprefix
).

Using `env={"NO_COLOR":"1"}` in subprocess does work, but it only stops coloring within what Python can control, like the `sqlite>` prompt. To disable color in GNU Readline itself, we need to run `readline.parse_and_bind("set colored-completion-prefix off")`.

picnixz replied:
Oh interesting. Maybe that's something to consider when we add colors?

cc @hugovk

tanloong committed:
Fix `run_pty()` doesn't return and test hangs

StanFromIreland commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,47 @@ def test_color(self):
                           '\x1b[35mnear "sel": syntax error\x1b[0m', err)
 
+@requires_subprocess()
+class CompletionTest(unittest.TestCase):
```
```suggestion
@force_class_not_colorized
@requires_subprocess()
class CompletionTest(unittest.TestCase):
```
This will stop all color for all tests. (Like all other classes in the file.)

tanloong replied:
Not work. This disables color in main process (the traceback becomes uncolored), but `run_pty()` still gives colored output containing control sequence. `run_pty()` runs in a subprocess.
![图片](https://github.com/user-attachments/assets/423f756a-0bb3-4fb5-b4dc-bcc02dfc0c14)

StanFromIreland replied:
Can we not then use `run_cli` (defined in the previous class) to make this much simpler overall?

StanFromIreland commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,47 @@ def test_color(self):
+    def setUpClass(cls):
+        readline = import_module("readline")
+        if readline.backend == "editline":
+            raise unittest.SkipTest("libedit readline is not supported")
```
```suggestion
            self.unittest.SkipTest("libedit readline is not supported")
```

tanloong replied:
You mean `unittest.TestCase.skipTest()`. There are [many uses of it](https://github.com/search?q=repo%3Apython%2Fcpython+%2F%28%3F-i%29%5C.skipTest%5C%28%2F+path%3A*.py&type=code) in the codebase, but there are also [quite a few uses of `raise unittest.SkipTest()`](https://github.com/search?q=repo%3Apython%2Fcpython+%2F%28%3F-i%29raise+unittest%5C.SkipTest%5C%28%2F+path%3A*.py&type=code), for example in [test_pdb.py#L4759](https://github.com/python/cpython/blob/main/Lib/test/test_pdb.py#L4759) which is used under exactly the same occasion (I actually copied that block from pdb to here). I think raising is OK?

picnixz replied:
It's incorrect to use `self.skipTest`. We're in a class method so we can only raise the exception. `skipTest` is a regular method, not a class method.

StanFromIreland commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,47 @@ def test_color(self):
+            raise unittest.SkipTest("libedit readline is not supported")
+
+    def test_keyword_completion(self):
+        script = "from sqlite3.__main__ import main; main()"
```
Why not just use run_cli from the interactiveconsole tests above?

tanloong replied:
With `run_cli()` both the `output` and `err` contain no completion candidates. I failed with `run_cli()` so I switched to `run_pty()` when writing the tests.
![图片](https://github.com/user-attachments/assets/e2bb9f3b-6e2e-4641-b627-f3685dd1f682)

StanFromIreland replied:
Then where are the completion candidates sent? The commands should be executed.

tanloong committed:
Revert "Remove color handling of output; If CI fails might need to add back"

This reverts commit 9d0373052b4da7f6c96f8c45984b22bab9e32fa5.

tanloong committed:
Turn off colored-completion-prefix for readline

tanloong committed:
No need to pass "NO_COLOR" to `run_pty()`

"NO_COLOR" only affects color for the `sqlite>` prompt, it does not help
for disabling color of completion prefix.

tanloong committed:
Flip name

tanloong committed:
Triggering completion on Ubuntu requires 2 tabs

tanloong committed:
Move KEYWORDS to C

tanloong committed:
Improve style of C code

4-space indents; 79 line width; outermost curly braces in column 1 in
function definition; blank line after local variable declaration;

tanloong committed:
Improve tests

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -70,6 +70,11 @@ Summary --- release highlights
+-------
+
+* Support keyword completion in the :mod:`sqlite3` command-line interface.
+  (Contributed by Tan Long in :gh:`133393`.)
```
Let's move this under the "Improved modules" section instead of "New features". The "New features" is reserved for PEP-sized items.

picnixz commented on Lib/sqlite3/_completer.py:
```diff
@@ -0,0 +1,36 @@
+def _complete(text, state):
+    global _completion_matches
+    if state == 0:
+        _completion_matches = [c + " " for c in SQLITE_KEYWORDS if c.startswith(text.upper())]
```
```suggestion
    if state == 0:
        text_upper = text.upper()
        _completion_matches = [c for c in SQLITE_KEYWORDS if c.startswith(text_upper)]
```

Otherwise, I *think* `text.upper()` is going to be called for each `c`. Then you can do `return _completion_matches[state] + " "` instead (no need to store the extra whitespace) as it would create a new string while here you can just create a new reference to an existing string.

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,74 @@ def test_color(self):
             self.assertIn('\x1b[1;35mOperationalError (SQLITE_ERROR)\x1b[0m: '
                           '\x1b[35mnear "sel": syntax error\x1b[0m', err)
 
+@requires_subprocess()
```
```suggestion

@requires_subprocess()
```

Two blank lines to separate classes.

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,74 @@ def test_color(self):
+            import readline
+            readline.parse_and_bind("set colored-completion-prefix off")
+            from sqlite3.__main__ import main; main()
+        """)
```
You can have a helper method:

```py
def write_input(self, input):
    script = textwrap.dedent("""
        import readline
        readline.parse_and_bind("set colored-completion-prefix off")
        from sqlite3.__main__ import main; main()
    """)
    return run_pty(script, input)
```

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,74 @@ def test_color(self):
+        output = run_pty(script, input, env={"NO_COLOR": "1"})
+        output_lines = output.decode().splitlines()
+        line_num = next(i for i, line in enumerate(output_lines, 1)
+                        if line.startswith(f"{self.PS1}xyzzy"))
```
```suggestion
        line_num = next(i for i, line in enumerate(output_lines, 1)
                        if line.startswith(f"{self.PS1}xyzzy"), -1)
```

Just to prevent some possible StopIteration if the test fails (easier to check that line_num == -1 indicating an incorrect line number)

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,74 @@ def test_color(self):
+        output = run_pty(script, input, env={"NO_COLOR": "1"})
+        output_lines = output.decode().splitlines()
+        candidates = []
+        for line in output_lines[-2::-1]:
```
Why are we iterating in reverse order and then sorting the candidates in reverse order as well?

tanloong replied:
Reversely iterating `output_lines` need one less line of code.

Candidates appear between the two lines that start with `self.PS1`.  The banner in the first 5 lines might change and occupy different number of lines in the future, if we iterate from top to bottom, to be reliable, we first need to find the index of the first `self.PS1`, then collect lines until the next `self.PS1`. For example:

```python
candidate_start = next(i for i, line in enumerate(output_lines) if line.startswith(self.PS1)) + 1
```

But if we iterate from the bottom up, we can skip finding the first `self.PS1`. Just start collecting lines as we go up until we encounter the first `self.PS1`.

<details>

<summary>
sqlite3 shell output example
</summary>

```
$ python -m sqlite3
sqlite3 shell, running on SQLite version 3.49.2
Connected to a transient in-memory database

Each command will be run using execute() on the cursor.
Type ".help" for more information; type ".quit" or CTRL-D to quit.
sqlite> # press tab
ABORT   # candidate start
ACTION
...     # many other candidates
WITH
WITHOUT # candidate end
sqlite>
```

</details>

But if you think this causes lower readability I am happy to modify it to iterate from top to bottom.

picnixz replied:
I think it'll be easier to just find the two lines that contain the prompt and take the slice of it.

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,74 @@ def test_color(self):
+            if line.startswith(self.PS1):
+                break
+            candidates.append(line.strip())
+        self.assertEqual(sorted(candidates, reverse=True), candidates)
```
I'm confused here. Are we checking that candidates are already sorted? or maybe you wanted to check against `SQLITE_KEYWORDS`?

tanloong replied:
I am thinking that in the future there could be more completion sources other than keywords, e.g. dot commands. Instead of `self.assertEqual(candidates, SQLITE_KEYWORDS)`, checking candidates against a sorted version of itself can avoid the bother of modifying the code when new sources are added. Maybe we need to additionally assert that the candidates contains all words from `SQLITE_KEYWORDS`?

```python
self.assertTrue(set(SQLITE_KEYWORDS).issubset(set(candidates)))
```

I forget to remove the unused line `from sqlite3._completer import SQLITE_KEYWORDS`. I will leave that line there for now in case we will add the above `issubset()` test.

picnixz replied:
> I am thinking that in the future there could be more completion sources other than keywords

This is a worry we'll need to address later. For now, let's assert exactly what we expect. Unless you have a future work planned for the next month about it.

tanloong replied:
OK I have pushed the update.

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -1,14 +1,19 @@
+import textwrap
 import unittest
 
+from _sqlite3 import SQLITE_KEYWORDS
```
When are we using it?

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,74 @@ def test_color(self):
+                break
+            candidates.append(line.strip())
+        self.assertEqual(sorted(candidates, reverse=True), candidates)
+
 if __name__ == "__main__":
```
```suggestion


if __name__ == "__main__":
```

picnixz commented on Modules/_sqlite/module.c:
```diff
@@ -404,6 +404,54 @@ pysqlite_error_name(int rc)
+add_sequence_constants(PyObject *module)
+{
+    PyObject *kwd;
+    const char *_keywords[] = {
```
When I said "we need it in C", I meant, we need the list of keywords given by sqlite3 directly, not by hardcoding it in C. Hardcoding it in C doesn't help.

tanloong committed:
Address Bénédikt's review

tanloong committed:
Revert "Improve style of C code"

This reverts commit 121b06913d9131347fe11231a4d5ff83c06e8dc4.

tanloong committed:
Revert "Move KEYWORDS to C"

This reverts commit 231b9e7a6eb98cd52fd7a25775579cb8783ddad0.

encukou committed:
Read keyword names dynamically

tanloong committed:
Check candidates against KEYWORDS

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -71,7 +71,6 @@ New features
 ============
 
 
-
```
Let's keep this blank line

picnixz reviewed (commented):
You can yourself to ACKS if it's not already done.

tanloong committed:
Use slice to get candidates

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,75 @@ def test_color(self):
+        slices = tuple(i for i, line in enumerate(output_lines) if line.startswith(self.PS1))
+        self.assertEqual(len(slices), 2)
+        start, end = slices
+        candidates = [c.strip() for c in output_lines[start+1 : end]]
```
```suggestion
        indices = [i for i, line in enumerate(output_lines)
                   if line.startswith(self.PS1)]
        self.assertEqual(len(indices), 2)
        start, end = indices[0] + 1, indices[1]
        candidates = [c.strip() for c in output_lines[start:end]]
```

I personally always find slice notation with an arithmetic operation inside unreadable due to spaces around `:` due to PEP-8. So I always try to have `i:j` instead of `i+X : j`. It's really a personal matter though but it's hard on me to read. I would have used `list(map(str.strip, output_lines[start:end]))` as well, but I know that pure-functional code is less readable to others.

tanloong replied:
I would have used `map()` as well if I thought of `str.strip` instead of the long `lambda c:c.strip()`. I think `map()` doesn't reduce too much readability but it make the code a bit cleaner.

Found that the `candidates` should not be sorted in `self.assertEqual()`, otherwise the test doesn't achieve its purpose.

erlend-aasland replied:
IMO, this is a misuse of PEP-8. Readability counts. Always. I strongly prefer @tanloong's original code, and I'm going to revert to that before landing.

tanloong committed:
Address Bénédikt's review

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,76 @@ def test_color(self):
+        self.assertEqual(len(indices), 2)
+        start, end = indices[0] + 1, indices[1]
+        candidates = list(map(str.strip, output_lines[start:end]))
+        self.assertEqual(candidates, list(KEYWORDS))
```
```suggestion
        candidates = tuple(map(str.strip, output_lines[start:end]))
        self.assertEqual(candidates, KEYWORDS)
```

tanloong committed:
Make candidates tuple

picnixz reviewed (approved):
On my side it looks great but I'll also want @erlend-aasland to have a look as he's the main maintainer. Maybe he'l ask for some stylistic changes (in which case I'm sorry if I suggested something wrong).

StanFromIreland commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,76 @@ def test_color(self):
+
+@requires_subprocess()
+class CompletionTest(unittest.TestCase):
+    PS1 = "sqlite> "
```
This is no longer freely customizable by users via sys.ps1 ?

erlend-aasland replied:
Can you explain your remark. Do you want changes?

encukou commented on Lib/sqlite3/_completer.py:
```diff
@@ -1,15 +1,41 @@
+            "ROLLBACK", "ROW", "ROWS", "SAVEPOINT", "SELECT", "SET",
+            "TABLE", "TEMP", "TEMPORARY", "THEN", "TIES", "TO",
+            "TRANSACTION", "TRIGGER", "UNBOUNDED", "UNION", "UNIQUE",
+            "UPDATE", "USING", "VACUUM", "VALUES", "VIEW", "VIRTUAL",
```
According to the [docs](https://sqlite.org/c3ref/keyword_check.html), `VACUUM` might not be a keyword on some builds of SQLite.
Python should ask SQLite for the list.

Since you mentioned you're not comfortable with C, I took a stab at it; see https://github.com/tanloong/cpython/pull/2

tanloong replied:
Oh, thanks a lot for the work!

tanloong committed:
Revert "Revert "Move KEYWORDS to C""

This reverts commit 226ea9fba9029b3c29092fae40040f5b4632eba5.

tanloong committed:
Revert "Revert "Improve style of C code""

This reverts commit b40982ab311335380fea7e42fb4eef7d4d949e74.

tanloong committed:
Merge pull request #2 from encukou/sqlite3-cli-completion

Read keyword names dynamically

picnixz commented on Lib/test/test_sqlite3/test_cli.py:
```diff
@@ -200,5 +205,76 @@ def test_color(self):
+        self.assertEqual(len(indices), 2)
+        start, end = indices[0] + 1, indices[1]
+        candidates = tuple(map(str.strip, output_lines[start:end]))
+        self.assertEqual(candidates, KEYWORDS)
```
```suggestion
        self.assertEqual(candidates, SQLITE_KEYWORDS)
```

tanloong committed:
Fix 'KEYWORDS' not found

tanloong committed:
Sort keywords before checking the equality

GNU Readline always sorts completions before displaying them. There is a [rl_sort_completion_matches](https://git.savannah.gnu.org/cgit/readline.git/tree/complete.c#n411) in Readline's source code but it's not exposed as a config flag.

tanloong committed:
Fix comparing between tuple and list

tanloong committed:
Fix comparing between tuple and list

tanloong committed:
Rename 'test_completion_order' to 'test_completion_for_nothing'

serhiy-storchaka commented on Modules/_sqlite/module.c:
```diff
@@ -404,6 +405,42 @@ pysqlite_error_name(int rc)
+            goto error;
+        }
+    }
+    if (PyModule_Add(module, "SQLITE_KEYWORDS", keywords) < 0) {
```
`PyModule_Add()` "eats" keywords, so `Py_XDECREF` should not be called after error. Just return -1. Or even return the result of `PyModule_Add()`.

StanFromIreland replied:
cc @encukou  who implemented this I believe.

tanloong replied:
May I try to update the C code. As this PR is sent from my fork, it would be more convenient that I make the change. I will convert if I mess it up.

serhiy-storchaka commented on Modules/_sqlite/module.c:
```diff
@@ -404,6 +405,42 @@ pysqlite_error_name(int rc)
+        if (!kwd) {
+            goto error;
+        }
+        if (PyTuple_SetItem(keywords, i, kwd) < 0) {
```
`PyTuple_SetItem()` "eats" kwd, so Py_DECREF should not be called after error.

erlend-aasland replied:
FTR, we use `PyTuple_SET_ITEM` elsewhere in the `_sqlite` extension module.

tanloong committed:
Don't decrease reference for `PyModule_Add()` and `PyTuple_SetItem()`

serhiy-storchaka reviewed (approved):
LGTM.

encukou committed:
Merge branch 'main' into sqlite3-cli-completion

encukou reviewed (dismissed):
Thank you! This looks great.
Let's check the buildbots before merging, as readline can be quirky. (So can the buildbots: not are stable so don't be alarmed if there are failures.)

tanloong committed:
Add @force_not_colorized_test_class

Disable coloring in main process to see if it resolves buildbots failures

encukou committed:
Merge branch 'main' into sqlite3-cli-completion

tanloong committed:
Add two '\b\b'; Skip tests on FreeBSD

tanloong committed:
Amend skipping reason

erlend-aasland commented on Modules/_sqlite/module.c:
```diff
@@ -404,6 +405,38 @@ pysqlite_error_name(int rc)
+static int
+add_keyword_tuple(PyObject *module)
+{
+    int count = sqlite3_keyword_count();
```
This feature requires SQLite 3.24.0 or higher; we need to wrap the C parts with a pre-processor guard, and then ensure that the Python part of the change is correctly disabled for SQLite 3.23.x and older.

tanloong replied:
I tried to add the guard. The C guard is a mimic of a [similar line](https://github.com/python/cpython/blob/ea108baa43286aeba742823b4b67440fc6d5ec0d/Modules/_sqlite/module.c#L677). Can you take another look at it?

erlend-aasland replied:
Marking as resolved; discussion continues in https://github.com/python/cpython/pull/133393#discussion_r2129811004.

erlend-aasland commented on Modules/_sqlite/module.c:
```diff
@@ -702,6 +735,11 @@ module_exec(PyObject *module)
         goto error;
     }
 
+    /* Set the keyword tuple */
```
This comment does not add any value (nor does the other similar comments above it, but that's not for this PR). I'd prefer if we do not add comments that spells out in text what the code clearly does.

erlend-aasland replied:
Addressed with 370dd8b1dba22f499ae8bc4f0bc5f881c948cfd4

tanloong committed:
Remove comment 'set the keyword tuple'

tanloong committed:
Disable keyword completion for SQLite<3.24.0

tanloong committed:
Don't disable the whole completion in case there will be more completion
sources added which might be available for even old SQLite.

serhiy-storchaka commented on Modules/_sqlite/module.c:
```diff
@@ -408,6 +408,9 @@ pysqlite_error_name(int rc)
 static int
 add_keyword_tuple(PyObject *module)
 {
+    if (sqlite3_libversion_number() < 3024000) {
```
We need to use compile-time check, otherwise the code will not compile.
```
#if SQLITE_VERSION_NUMBER >= 3024000
```

erlend-aasland replied:
Marking as resolved; discussion continues in https://github.com/python/cpython/pull/133393#discussion_r2129811004.

erlend-aasland replied:
Resolved with 588fb6a69a383786dc8221b733ea66e4d0dc9363

tanloong committed:
Use compile-time check

tanloong committed:
Correct #if usage

serhiy-storchaka commented on Modules/_sqlite/module.c:
```diff
@@ -404,6 +405,41 @@ pysqlite_error_name(int rc)
+{
+    #if SQLITE_VERSION_NUMBER < 3024000
+        return 0;
+    #endif
```
It does not work this way. If SQLITE_VERSION_NUMBER < 3024000, you should not use `sqlite3_keyword_count` in any way, even in a code that is not executed at run time. It is undefined.

Wrap the whole `add_keyword_tuple()` definition and its call site in `#if`/`#endif`.

erlend-aasland replied:
Marking as resolved; discussion continues in https://github.com/python/cpython/pull/133393#discussion_r2129811004.

erlend-aasland replied:
Resolved with 588fb6a69a383786dc8221b733ea66e4d0dc9363

tanloong committed:
Wrap add_keyword_tuple() definition and its call in #if/#endif

serhiy-storchaka commented on Modules/_sqlite/module.c:
```diff
@@ -404,6 +405,40 @@ pysqlite_error_name(int rc)
     return NULL;
 }
 
+#if SQLITE_VERSION_NUMBER < 3024000
```
You need an opposite condition.

Note also that the tests failed. You need them be passed when build with old Sqlite. Since SQLITE_VERSION_NUMBER is not exposed, you have to rely on the result of `hasattr(_sqlite3, 'SQLITE_KEYWORDS')` to determine the expected behavior.

erlend-aasland replied:
Resolved with 588fb6a69a383786dc8221b733ea66e4d0dc9363

erlend-aasland committed:
Suggestions to python/cpython#133393

tanloong committed:
Merge pull request #3 from erlend-aasland/suggestion

Suggestions to python/cpython#133393

erlend-aasland committed:
Merge branch 'main' into sqlite3-cli-completion

erlend-aasland commented on Modules/_sqlite/module.c:
```diff
@@ -404,6 +405,40 @@ pysqlite_error_name(int rc)
+    for (int i = 0; i < count; i++) {
+        const char *keyword;
+        int size;
+        int result = sqlite3_keyword_name(i, &keyword, &size);
```
For the record:
We _could_ just `assert` that the result is `SQLITE_OK`. This is just an out-of-bounds check, and we know that we are within bounds here. OTOH, we don't know how the SQLite internals may change, so let's keep it like it is.

erlend-aasland commented on Doc/whatsnew/3.15.rst:
```diff
@@ -134,6 +134,13 @@ shelve
+sqlite3
+-------
+
+* Support keyword completion in the :mod:`sqlite3` command-line interface.
```
Let's be explicit:

```suggestion
* Support SQL keyword completion in the :mod:`sqlite3` command-line interface.
```

erlend-aasland committed:
Update Doc/whatsnew/3.15.rst

erlend-aasland committed:
Merge branch 'main' into sqlite3-cli-completion

erlend-aasland commented on Lib/sqlite3/_completer.py:
```diff
@@ -0,0 +1,42 @@
+            # libedit uses "^I" instead of "tab"
+            command_string = "bind ^I rl_complete"
+        else:
+            command_string = "tab: complete"
```
Are we sure we're not missing anything for readline? I can't get the completion query (the yes/no-thingy) to work on readline.

erlend-aasland replied:
`"set show-all-if-unmodified on"` does not make an impact. It only results in everything (including the prompt) being shown after the first tab press.

tanloong replied:
I copied this command_string part from pdb.py https://github.com/python/cpython/blob/main/Lib%2Fpdb.py

I think this code is correct, otherwise developers on pdb side should have spotted the problem when testing pdb.

erlend-aasland replied:
I suspect it may be a problem with my readline installation.

tanloong replied:
I've just returned home and can reproduce this on my machine. There is a [same issue](https://bugs.python.org/issue22923) years ago, the problem is that  [libreadline](https://cgit.git.savannah.gnu.org/cgit/readline.git/tree/complete.c#n541) always set the answer as 'yes'.
